proposals/0000-fine-grained-unused-warnings.rst

```diff
@@ -0,0 +1,205 @@
+for the unused binding (in this case ``-Wunused-top-binds``) and the the one for
+the transitively unused binding itself (in this case ``-Wunused-imports``) are
+enabled.
+
+If ``foo`` is only used in a situation like
+
+::
+
+  bar = baz + 2
+    where baz = foo * 2
+	
+and bar is unused, a question that arises is whether the warning about ``foo`` should reference ``bar``, or ``baz``.
+
+In this particular case, since there would be no warning about ``baz`` being unused, it seem more helpful if the warning references ``bar``.
+
+In general, a warning should only reference other bindings about which there is an "unused" warning.
+
+In practice, that typically means warnings about transitively unused top-level bindings will tend to reference other top-level bindings, and warnings about transitively unused
+local bindings will reference other local bindings. However, if ``baz`` were unused, it would be referenced in the warning instead, since using ``bar`` in that case would not get rid of the warning about ``foo``.
+
+If a binding ``b`` is only used recursively, the warning message will say that it is only used in ``b``.
+
+If a group of bindings ``b1``, ..., ``bn`` is mutually recursive, and none of
+the bindings are otherwise used, each of the bindings will be treated as transitively unused.
+In this case, the binding site will reference the bindings of this group in whose definitions it directly occurs, e.g.
+
+::
+
+  Foo.hs:6:1: warning: [-Wunused-top-binds]
+      ‘b1’ is defined but only used in the following unused bindings: ‘b2’, ‘b4’
+
+Related warnings that are not affected by this proposal are ``-Wunused-pattern-bindings``,
+``-Wunused-packages``, and ``-Wunused-do-bind``, since these are not about binding names.
+
+Examples
+--------
+
+NB: The only difference between these examples and how these warnings are currently presented is that the warnings about transitively unused bindings currently don't list the bindings in which they occur, and have the same phrasing as regular unused warnings - with the exception of the unused import warning and unused forall warnings, which don't currently occur if they are only transitively unused.
+
+::
+
+  module Foo () where
+
+  import Data.List as L
+
+  foo = L.intercalate bar
+
+  bar = baz
+    where baz = undefined
+          quux = wibble
+          wibble = worble
+          worble = quux
+        
+  far :: forall a (b :: a) c . c
+  far = far
+
+::
+
+  Foo.hs:3:1: warning: [-Wunused-imports, -freport-indirectly-used-bindings]:
+      The import of ‘Data.List’ is only used by the following unused binding: ‘foo’
+        except perhaps to import instances from ‘Data.List’
+      To import instances alone, use: import Data.List()
+
+  Foo.hs:5:1: warning: [-Wunused-top-binds]
+      Defined but not used: ‘foo’
+
+  Foo.hs:7:1: warning: [-Wunused-top-binds]
+      ‘bar' is defined but only used in the following unused binding: `foo’
```

Comment:
Being pedantic, almost all occurrences of "only used" in the proposal should be "used only".